src/sock.c

```diff
@@ -260,11 +260,12 @@ static long mg_sock_recv(struct mg_connection *c, void *buf, size_t len) {
 static long read_conn(struct mg_connection *c) {
   long n = -1;
   if (c->recv.len >= MG_MAX_RECV_BUF_SIZE) {
-    mg_error(c, "max_recv_buf_size reached");
+    return n;
   } else if (c->recv.size - c->recv.len < MG_IO_SIZE &&
+             c->recv.size + MG_IO_SIZE <= MG_MAX_RECV_BUF_SIZE &&
              !mg_iobuf_resize(&c->recv, c->recv.size + MG_IO_SIZE)) {
     mg_error(c, "oom");
-  } else {
+  } else if (c->recv.size > c->recv.len) {
```

Comment:
I added this which appears to be sufficient to prevent reads when the buffer is full.